File: cs50/pset6/credit.py
from cs50 import get_int
# initialize variables, to include card number, digit count, and sums for algorithm
ccnum = get_int("Number: ")
check = ccnum
count = 0
summed = 0
p = []
# loop to determine how many digits are in the number
while check != 0:
    pnext = check % 10
    p.append(pnext)
    check = (check - p[count]) / 10
    count += 1
# add an empty element in case loop extends beyond size of array since number could be even or odd
p.append(0)
# do algo math, which goes between alternate digits of the number
for i in range(0, count + 1, 2):
    summed = summed + p[i]
for i in range(1, count + 1, 2):
    if 2 * p[i] > 9:
        summed += 2 * p[i] - 9
    else:
        summed += 2 * p[i]
    #do checks to determine which card we have and print the results
if summed % 10 == 0:
    if p[count - 1] == 4 and (count == 16 or count == 13):
        print("VISA")
    elif count == 15 and p[count - 1] == 3 and (p[count - 2] == 4 or p[count - 2] == 7):
        print("AMEX")
    elif count == 16 and p[count - 1] == 5 and (p[count - 2] == 1 or p[count - 2] == 2 or p[count - 2] == 3 or p[count - 2] == 4 or p[count - 2] == 5):
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")

File: cs50/pset1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long ccnum = get_long("Number: ");
    long check = ccnum;
    int count = 0, sum = 0, p[33];
// Find how many digits and what they are, store in array
    do
    {
        p[count] = check % 10;
        check = (check - p[count]) / 10;
        count++;
    }
    while (check != 0);
// Run checksum
// Starting with first digit and skipping up
    for (int j = 0; j < count; j = j + 2)
    {
        sum = sum + p[j];
    }
// Then for the second digit and skipping up, this time double it
    for (int j = 1; j < count + 1; j = j + 2)
    {
        if (2 * p[j] > 9)
        {
            // This considers the value of the doubled digit's digits added
            sum = sum + (2 * p[j]) - 9;
        }
        else
        {
            sum = sum + 2 * p[j];
        }
    }
// If checksum passes, see if a card matches and identify
    if (sum % 10 == 0)
    {
        if (p[count - 1] == 4 && (count == 16 || count == 13))
        {
            printf("VISA\n");
        }
        else if (count == 15 && p[count - 1] == 3 && (p[count - 2] == 4 || p[count - 2] == 7))
        {
            printf("AMEX\n");
        }
        else if (count == 16 && p[count - 1] == 5 && (p[count - 2] == 1 || p[count - 2] == 2 || p[count - 2] == 3 || p[count - 2] == 4
                 || p[count - 2] == 5))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
